wandb/sdk/launch/runner/gcp_vertex.py

```diff
@@ -0,0 +1,244 @@
+        container_workdir = container_inspect["ContainerConfig"].get("WorkingDir", "/")
+        container_env: List[str] = container_inspect["ContainerConfig"]["Env"]
+
+        if launch_project.docker_image is None or launch_project.build_image:
+            image_uri = construct_gcp_image_uri(
+                launch_project, gcp_artifact_repo, gcp_project, gcp_docker_host,
+            )
+            image = build_docker_image_if_needed(
+                launch_project=launch_project,
+                api=self._api,
+                copy_code=copy_code,
+                workdir=container_workdir,
+                container_env=container_env,
+                runner_type="gcp-vertex",
+                image_uri=image_uri,
+                command_args=command_args,
+            )
+        else:
+            image = launch_project.docker_image
+
+        docker_push(image)  # todo: when aws pr is merged, use docker python tooling
+
+        if self.backend_config.get("runQueueItemId"):
+            try:
+                self._api.ack_run_queue_item(
+                    self.backend_config["runQueueItemId"], launch_project.run_id
+                )
+            except CommError:
+                wandb.termerror(
+                    "Error acking run queue item. Item lease may have ended or another process may have acked it."
+                )
+                return None
+
+        job = aiplatform.CustomContainerTrainingJob(
+            display_name=gcp_training_job_name, container_uri=image,
+        )
+        submitted_run = VertexSubmittedRun(job)
+
+        # todo: support gcp dataset?
+
+        wandb.termlog(
+            "Running training job {name} on {compute}.".format(
+                name=gcp_training_job_name, compute=gcp_machine_type
+            )
+        )
+
+        # when sync is True, vertex blocks the main thread on job completion. when False, vertex returns a Future
+        # on this thread but continues to block the process on another thread. always set sync=False so we can get
+        # the job info (dependent on job._gca_resource)
+        job.run(
+            machine_type=gcp_machine_type,
+            accelerator_type=gcp_accelerator_type,
+            accelerator_count=gcp_accelerator_count,
+            replica_count=1,
+            sync=False,
+        )
+        while job._gca_resource is None:
+            # give time for the gcp job object to be created, this should only loop a couple times max
+            time.sleep(1)
+
+        wandb.termlog(
+            "View your job status and logs at {url}.".format(
+                url=submitted_run.get_page_link()
+            )
+        )
+
+        # hacky: if user doesn't want blocking behavior, kill both main thread and the background thread. job continues
+        # to run remotely. this obviously doesn't work if we need to do some sort of postprocessing after this run fn
+        if not synchronous:
+            os._exit(0)
```

Comment:
Couldn't we just kill the background thread and keep the submitted_run job object for the agent to use and report on?

Reply:
discussed offline: this doesn't work because the background thread is controlled in the vertex api and we don't have access to it